clippy_lints/src/extra_unused_type_parameters.rs

```diff
@@ -0,0 +1,118 @@
+use clippy_utils::diagnostics::span_lint;
+use clippy_utils::trait_ref_of_method;
+use rustc_data_structures::fx::FxHashMap;
+use rustc_hir::intravisit::{walk_impl_item, walk_item, walk_param_bound, walk_ty, Visitor};
+use rustc_hir::{GenericParamKind, Generics, ImplItem, ImplItemKind, Item, ItemKind, Ty, TyKind, WherePredicate};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_middle::hir::nested_filter;
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+use rustc_span::{def_id::DefId, Span};
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for type parameters in generics that are never used anywhere else.
+    ///
+    /// ### Why is this bad?
+    /// Functions cannot infer the value of unused type parameters; therefore, calling them
+    /// requires using a turbofish, which serves no purpose but to satisfy the compiler.
+    ///
+    /// ### Example
+    /// ```rust
+    /// // unused type parameters
+    /// fn unused_ty<T>(x: u8) {
+    ///     // ..
+    /// }
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// fn no_unused_ty(x: u8) {
+    ///     // ..
+    /// }
+    /// ```
+    #[clippy::version = "1.67.0"]
+    pub EXTRA_UNUSED_TYPE_PARAMETERS,
+    complexity,
+    "unused type parameters in function definitions"
+}
+declare_lint_pass!(ExtraUnusedTypeParameters => [EXTRA_UNUSED_TYPE_PARAMETERS]);
+
+struct TypeWalker<'cx, 'tcx> {
+    cx: &'cx LateContext<'tcx>,
+    map: FxHashMap<DefId, Span>,
+}
+
+impl<'cx, 'tcx> TypeWalker<'cx, 'tcx> {
+    fn new(cx: &'cx LateContext<'tcx>, generics: &Generics<'tcx>) -> Self {
+        Self {
+            cx,
+            map: generics
+                .params
+                .iter()
+                .filter_map(|param| match param.kind {
+                    GenericParamKind::Type { .. } => Some((param.def_id.into(), param.span)),
+                    _ => None,
+                })
+                .collect(),
+        }
+    }
+
+    fn emit_lint(&self) {
+        for span in self.map.values() {
+            span_lint(
```

Comment:
This should be a `span_lint_and_sugg`. However, I can see that producing a suggestion that suggests to remove the parameter might be hard. (how does `extra_unused_lifetimes` do this?). But this should at least be a `span_lint_and_help`.

Reply:
Actually, `extra_unused_lifetimes` only uses `span_lint`. I think part of the challenge is in determining the span for the help/suggestion, as it's possible that _all_ the generic parameters on a function go unused, in which case the fix requires also removing the surrounding angle brackets. Or, if there are still some parameters left over, the angle brackets remain. I could certainly add a help string and keep the help span `None`.

Also, what's your opinion on combining multiple unused parameters into a single lint message? I have a test case for this: `unused_multi<T, U>(x: u8) {}`

Reply:
Yes, figuring out spans here will be the hardest part. Also figuring out what `,` to remove together with the generic parameter will be hard.

Also this is Rust, so `fn foo<>() {}` is [totally valid syntax](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=b320f376223a5c58a50e2e5203c79321) which will then be removed by `rustfmt`. So you don't have to worry about the brackets. 

So when figuring out the spans, you will only have to care about the bounds and the `,`s. 

But this is all extra credit work. Producing cool suggestions can be fun though :) 

---

> Also, what's your opinion on combining multiple unused parameters into a single lint message?

Also a cool thing to do. You will then have to suggest things with `multipart_suggestion`, which is not `rustfix`able, though.

If you're motivated to try out things like this, go ahead and do it. You can post the results to Zulip and we can figure out what looks and feels best. However, I would already be fine with merging this with just a `span_lint_and_help`.

Reply:
```rust
pub fn foo<T,>() {}
```

is also valid. So you probably want a span until the next `,` or `>`, if no `,` exists.

Take a look at `SourceMap` and its functions to produce such spans: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_span/source_map/struct.SourceMap.html#method.span_until_char

Reply:
I've pushed my attempt at doing this; let me know what you think.